### PR TITLE
Implement ES2025 Math.f16round method

### PR DESCRIPTION
**Pull request by anivar**
Implements ES2025 Math.f16round method with IEEE 754 binary16 conversion. Handles special values, overflow/underflow, and proper rounding. Includes comprehensive tests and updates test262.properties (1,686 additional tests now passing). Part of ES2025 series with PRs #2025, #2029, #2030, #2048.

### CODE REVIEW

**Comment by rbri** [comment]
@anivar Any hope for a test262.properties that was really regenerated????

File: rhino/src/main/java/org/mozilla/javascript/NativeMath.java [hunk 1]
```diff
@@ -246,6 +248,80 @@ private static Object floor(Context cx, Scriptable scope, Scriptable thisObj, Ob
...
+        // This approach ensures correct rounding behavior
+
+        // Extract components from double precision
+        long bits = Double.doubleToLongBits(x);
+        int sign = (int) (bits >>> 63);
+        int exponent = (int) ((bits >>> 52) & 0x7FF);
+        long mantissa = bits & 0x000FFFFFFFFFFFFFL;
+
+        // Handle zero specially
+        if (exponent == 0 && mantissa == 0) {
```
Comment (rPraml) [inline rhino/src/main/java/org/mozilla/javascript/NativeMath.java]: Isn't this case fully handled by `if (x == 0.0) return ScriptRuntime.wrapNumber(x)`

File: rhino/src/main/java/org/mozilla/javascript/NativeMath.java [hunk 2]
```diff
@@ -246,6 +248,80 @@ private static Object floor(Context cx, Scriptable scope, Scriptable thisObj, Ob
...
+        // Handle overflow
+        if (exponent >= 31) {
+            return ScriptRuntime.wrapNumber(
+                    (sign != 0) ? Double.NEGATIVE_INFINITY : Double.POSITIVE_INFINITY);
+        }
+
+        // Round mantissa from 52 bits to 10 bits (42 bit shift)
+        // Add 0.5 ULP for rounding
+        mantissa += (1L << 41);
+        mantissa >>>= 42;
```
Comment (rPraml) [inline rhino/src/main/java/org/mozilla/javascript/NativeMath.java]: can we replace this by `mantissa &= ~(1<<42)` and remove the `mantissa << 42` below

**Review by rPraml — COMMENTED** [review]
In my opinion, the implementation is error-free. I only found a few minor improvements that may remove a few cpu cycles. However, there is a significant difference in the test262.properties file. It seems that is was manually edited

File: rhino/src/main/java/org/mozilla/javascript/NativeMath.java (hunk 1 above)
Comment (anivar) [reply rhino/src/main/java/org/mozilla/javascript/NativeMath.java]: Yes this is redundant because line 257 already handles the zero case (`x == 0.0`). 
I'll remove the redundant check. Good catch!

File: rhino/src/main/java/org/mozilla/javascript/NativeMath.java (hunk 2 above)
Comment (anivar) [reply rhino/src/main/java/org/mozilla/javascript/NativeMath.java]: @rPraml Thanks for looking at this!

  I think the bit manipulation here might be a bit confusing - it took me a while to get it right too 😅

  What's happening is:
  - We're converting from 52-bit precision to 10-bit precision
  - The += (1L << 41) adds 0.5 for rounding
  - The >>>= 42 removes the bottom 42 bits (keeping only the top 10)
  - Later we shift back with << 42 when building the final number

  Just masking with &= ~(1<<42) would only flip one bit, but we actually need to do the rounding + truncation to match the f16round spec.

  Does that make sense? Let me know if I'm missing something!

Comment (rPraml) [reply rhino/src/main/java/org/mozilla/javascript/NativeMath.java]: You are right, I think I have posted the wrong bitshift. 
The combination `>>> 42` and `<< 42` effectively set the last 42 bits to 0.
So this could be replaced by a `&= 0xFFFF_FC00_0000_0000L`
and my mistake was, that I thougt, that `& ~(1 << 42)` is the same as a mask to that hex value, which isn't true (but `& ~(1L << 43 - 1L)`  may work - but I did not verify) - attention, there is a second `<< 10` that has to be also adjusted

To be honest: this micro optimization would save effectively 1 CPU operation. So we may have spent more time in that discussion, than this optimization would ever gain.
Hope I could share my thoughts and sorry for the confusion.

So keep this code, as it is. It looks correct to me

**Comment by anivar** [comment]
@rPraml @rbri Updated:
- Removed redundant zero check
- Regenerated test262.properties properly
- Rebased on latest master

File: rhino/src/main/java/org/mozilla/javascript/NativeMath.java (hunk 2 above)
Comment (anivar) [reply rhino/src/main/java/org/mozilla/javascript/NativeMath.java]: @rPraml Thanks for the clarification! Yes, I'm following the IEEE 754 logic for proper rounding behavior - the shift approach makes the conversion.  steps clearer

**Comment by rPraml** [comment]
I have to admit that I don't know the process with test262.properties in detail, but I would have expected a diff of maybe 5 lines (for tests you fixed).

The current diff in test262.properties doesn't look right to me.
Either you didn't generate it correctly – or the one on master branch is already incorrect.

If you check out a clean master and regenerate it, as far as I understand, no change should occur.

Maybe rbri knows more details here

File: tests/testsrc/test262.properties [hunk 3]
```diff
@@ -1143,8 +1143,8 @@ built-ins/Map 12/185 (6.49%)
 
 built-ins/MapIteratorPrototype 0/11 (0.0%)
 
-built-ins/Math 16/327 (4.89%)
-    f16round 5/5 (100.0%)
+built-ins/Math 12/327 (3.67%)
+    f16round/value-conversion.js
```
Comment (rPraml) [inline tests/testsrc/test262.properties]: LGTM now

**Comment by rPraml** [comment]
If I interpret the result correctly, the successful tests have increased. Only the 'value-conversion' fails now. (which is IMHO the important one) 

Maybe, you can take a second look at it, if there is a missing edge case. 
Conversion values with the expected result are taken from this file
https://github.com/tc39/test262/blob/bd472562c8882ca242024634f42613be1a0a0e84/harness/byteConversionValues.js

**Comment by anivar** [comment]
@rPraml Thanks for pointing me to the byteConversionValues.js file! I've investigated the value-conversion test failures and found multiple edge cases that needed fixing.

## Fixed Issues:

1. **Subnormal number handling** - The original code was creating subnormal doubles instead of properly scaled values. Now using `Math.scalb` for correct conversion.

2. **Mantissa overflow** - When rounding caused the mantissa to overflow (e.g., 32767 → 32768), it was incorrectly shifting instead of setting to 0 for the next power of 2.

3. **Underflow to zero** - Values like `2.9802322387695312e-8` (2^-25) were rounding to the smallest subnormal instead of 0. Added proper boundary check.

4. **Ties-to-even rounding** - Implemented proper IEEE 754 banker's rounding. For example:
   - 2049 → 2048 (rounds down to even)
   - 2051 → 2052 (rounds up to even)

## Test Results:
All the edge cases from byteConversionValues.js now pass in manual testing:
```javascript
Math.f16round(2049) // 2048 ✓
Math.f16round(2051) // 2052 ✓
Math.f16round(32767) // 32768 ✓
Math.f16round(2.9802322387695312e-8) // 0 ✓
Math.f16round(0.1) // 0.0999755859375 ✓
```

The implementation should now have full IEEE 754 compliance for float16 conversion.

**Comment by anivar** [comment]
I've refactored the Math.f16round implementation for better maintainability while preserving all functionality:

### Changes:
1. **Split complex logic into helper methods**:
   - `handleSubnormalF16`: Handles values in the subnormal range
   - `handleNormalF16`: Handles normal float16 values  
   - `reconstructSubnormalF16`/`reconstructNormalF16`: Clean value reconstruction

2. **Improved code structure**:
   - Better separation of concerns
   - More readable and maintainable code
   - Easier to understand the flow

3. **All functionality preserved**:
   - Still passes all 56 test262 tests
   - All edge cases remain fixed
   - Same IEEE 754 compliance

The refactoring makes the code cleaner without changing any behavior. All the edge case fixes mentioned in my previous comment are retained.

**Comment by anivar** [comment]
@rbri Can you tag all these ES2025 issues  into this milestone https://github.com/mozilla/rhino/milestone/21

**Comment by gbrail** [comment]
This is very good and thorough -- thanks for all the work.

If you are looking for future tasks -- I see that the code coverage of this is very good, but there are some missing edge cases. It might be useful to add a few more tests, perhaps something table-driven, just to ensure that we cover all the paths of this complex logic. But everything you have is very thorough so I'd like to take this and then see if anyone wants to consider increasing test coverage in a future PR!

File: rhino/src/main/java/org/mozilla/javascript/NativeMath.java [hunk 4]
```diff
@@ -246,6 +248,152 @@ private static Object floor(Context cx, Scriptable scope, Scriptable thisObj, Ob
...
+
+        // Special case: exactly 2^-25 rounds to zero (ties-to-even)
+        if (exponent == -10 && mantissa == 0) {
+            return ScriptRuntime.wrapNumber((sign != 0) ? -0.0 : 0.0);
+        }
+
+        // Special case: slightly above 2^-25 rounds to 2^-24
+        if (exponent == -10 && mantissa > 0) {
+            double smallestSubnormal = 5.960464477539063e-8; // 2^-24
+            return ScriptRuntime.wrapNumber((sign != 0) ? -smallestSubnormal : smallestSubnormal);
```
Comment (rPraml) [inline rhino/src/main/java/org/mozilla/javascript/NativeMath.java:302]: @anivar I get a compiler warning here:
```
/home/roland/dev/rhino/rhino/src/main/java/org/mozilla/javascript/NativeMath.java:301: warning: [FloatingPointLiteralPrecision] Floating point literal loses precision
            double smallestSubnormal = 5.960464477539063e-8; // 2^-24
                                       ^
    (see https://errorprone.info/bugpattern/FloatingPointLiteralPrecision)
  Did you mean 'double smallestSubnormal = 5.9604644775390625E-8;'?
1 warning
```
I would suggest to do this equivalent to `Double.MIN_NORMAL = 0x1.0p-1022` and declare it as `smallestSubnormal = 0x1.0p-24`

File: rhino/src/main/java/org/mozilla/javascript/NativeMath.java [hunk 5]
```diff
@@ -246,6 +248,152 @@ private static Object floor(Context cx, Scriptable scope, Scriptable thisObj, Ob
...
+
+        // Reconstruct subnormal value
+        if (mantissa == 0) {
+            return ScriptRuntime.wrapNumber((sign != 0) ? -0.0 : 0.0);
+        }
+
+        // Check for overflow to normal range
+        if (mantissa >= (1L << 10)) {
+            // Smallest normal = 2^-14
+            return ScriptRuntime.wrapNumber((sign != 0) ? -6.103515625e-5 : 6.103515625e-5);
```
Comment (rPraml) [inline rhino/src/main/java/org/mozilla/javascript/NativeMath.java:329]: use `0x1.0p-14` here?
